pkg/syncer/controller/syncerconfigmanager.go

```diff
@@ -129,6 +132,99 @@ func (s *SyncerConfigManager) upsertNamespaceScoped(syncerConfig edgev1alpha1.Sy
 	s.syncConfigManager.upsert(edgeSyncConfig)
 }
 
+type typeFlatNamespacedObject struct {
+	APIVersion string
+	v1.GroupResource
+	Namespace string
+	Names     []string
+}
+
+type typeTableOfFlatNamespacedObject struct {
+	FlatNamespacedObjects []typeFlatNamespacedObject
+}
+
+func (t *typeTableOfFlatNamespacedObject) filter(predicate func(row typeFlatNamespacedObject) bool) *typeTableOfFlatNamespacedObject {
+	flatNamespacedObjects := []typeFlatNamespacedObject{}
+	for _, obj := range t.FlatNamespacedObjects {
+		if predicate(obj) {
+			flatNamespacedObjects = append(flatNamespacedObjects, obj)
+		}
+	}
+	return &typeTableOfFlatNamespacedObject{FlatNamespacedObjects: flatNamespacedObjects}
+}
+
+func (t *typeTableOfFlatNamespacedObject) getAllNamespaces() []string {
+	namespaces := sets.String{}
+	for _, obj := range t.FlatNamespacedObjects {
+		namespaces.Insert(obj.Namespace)
+	}
+	return namespaces.List()
+}
+
+func (s *SyncerConfigManager) upsertNamespacedObjects(syncerConfig edgev1alpha1.SyncerConfig, upstreamGroupResourcesList []*restmapper.APIGroupResources) {
+	flatNamespacedObjects := []typeFlatNamespacedObject{}
+	for _, nsObject := range syncerConfig.Spec.NamespacedObjects {
+		for _, obj := range nsObject.ObjectsByNamespace {
+			flatNamespacedObjects = append(flatNamespacedObjects, typeFlatNamespacedObject{
+				APIVersion:    nsObject.APIVersion,
+				GroupResource: nsObject.GroupResource,
+				Namespace:     obj.Namespace,
+				Names:         obj.Names,
+			})
+		}
+	}
+	namespacedObjectsTable := typeTableOfFlatNamespacedObject{FlatNamespacedObjects: flatNamespacedObjects}
+
+	requiredNamespaces := namespacedObjectsTable.filter(func(row typeFlatNamespacedObject) bool { return len(row.Names) > 0 }).getAllNamespaces()
+
+	s.logger.V(3).Info("upsert namespaced objects as syncerConfig to syncConfigManager stores", "syncerConfigName", syncerConfig.Name, "numNamespaces", len(requiredNamespaces))
+	if lgr := s.logger.V(4); lgr.Enabled() {
+		for _, agrs := range upstreamGroupResourcesList {
+			lgr.Info("APIGroupResources", "group", agrs.Group, "bar", agrs.VersionedResources)
+		}
+	}
+	edgeSyncConfigResources := []edgev1alpha1.EdgeSyncConfigResource{}
+	for _, namespace := range requiredNamespaces {
+		edgeSyncConfigResourceForNamespace := edgev1alpha1.EdgeSyncConfigResource{
+			Group:   "",
+			Version: "v1",
+			Kind:    "Namespace",
+			Name:    namespace,
+		}
+		edgeSyncConfigResources = append(edgeSyncConfigResources, edgeSyncConfigResourceForNamespace)
+	}
+
+	for _, object := range namespacedObjectsTable.FlatNamespacedObjects {
+		group := object.Group
+		version := object.APIVersion
+		resource := object.Resource
+		gvr := schema.GroupVersionResource{Group: group, Version: version, Resource: resource}
+		versionedResources := findVersionedResourcesByGVR(group, version, resource, upstreamGroupResourcesList, s.logger)
```

Comment:
Why does this func return a slice instead of a single match?  `resource` is never `"*"`.  Also, it might be worth logging a warning message when there is no match, something higher level might have expected one.

Reply:
Since `apiGroupResources.VersionedResources[version]` is array of `v1.APIResource` at https://github.com/kubestellar/kubestellar/blob/8731b55053a056a7f3211279ef6cbd3e502c2fce/pkg/syncer/controller/syncerconfigmanager.go#L355, I made a code to collect all elements of the array matching with `resource`.

>  it might be worth logging a warning message when there is no match, something higher level might have expected one.

I agree with it.

Reply:
My question is whether the findVersionedResourcesByGVR would ever return a slice holding more than one member.

I asked about the underlying question at https://kubernetes.slack.com/archives/C0EG7JC6T/p1694053099115029 .  As best I can tell, the answer there implies that findVersionedResourcesByGVR will never return a slice longer than 1.